presto-hive/src/main/java/com/facebook/presto/hive/parquet/predicate/ParquetPredicateUtils.java

```diff
@@ -73,88 +76,80 @@ public static boolean isStatisticsOverflow(Type type, ParquetIntegerStatistics p
+    private static Map<ColumnDescriptor, Statistics<?>> getStatistics(BlockMetaData blockMetadata, MessageType fileSchema, MessageType requestedSchema)
     {
-        ImmutableMap.Builder<Integer, Statistics<?>> statistics = ImmutableMap.builder();
-        for (int ordinal = 0; ordinal < blockMetadata.getColumns().size(); ordinal++) {
-            Statistics<?> columnStatistics = blockMetadata.getColumns().get(ordinal).getStatistics();
+        ImmutableMap.Builder<ColumnDescriptor, Statistics<?>> statistics = ImmutableMap.builder();
+        for (ColumnChunkMetaData columnMetaData : blockMetadata.getColumns()) {
+            Statistics<?> columnStatistics = columnMetaData.getStatistics();
             if (columnStatistics != null) {
-                statistics.put(ordinal, columnStatistics);
+                String[] paths = columnMetaData.getPath().toArray();
+                Optional<RichColumnDescriptor> descriptor = getDescriptor(fileSchema, requestedSchema, Arrays.asList(paths));
+                if (descriptor.isPresent()) {
+                    statistics.put(descriptor.get(), columnStatistics);
+                }
             }
         }
         return statistics.build();
     }
 
-    private static Map<Integer, ParquetDictionaryDescriptor> getDictionariesByColumnOrdinal(
-            BlockMetaData blockMetadata,
-            ParquetDataSource dataSource,
-            MessageType requestedSchema,
-            TupleDomain<HiveColumnHandle> effectivePredicate)
+    private static Map<ColumnDescriptor, ParquetDictionaryDescriptor> getDictionaries(BlockMetaData blockMetadata, ParquetDataSource dataSource, MessageType fileSchema, MessageType requestedSchema, TupleDomain<ColumnDescriptor> parquetTupleDomain)
     {
-        ImmutableMap.Builder<Integer, ParquetDictionaryDescriptor> dictionaries = ImmutableMap.builder();
-        for (int ordinal = 0; ordinal < blockMetadata.getColumns().size(); ordinal++) {
-            ColumnChunkMetaData columnChunkMetaData = blockMetadata.getColumns().get(ordinal);
-
-            for (int i = 0; i < requestedSchema.getColumns().size(); i++) {
-                ColumnDescriptor columnDescriptor = requestedSchema.getColumns().get(i);
-                if (isColumnPredicate(columnDescriptor, effectivePredicate) &&
-                        columnChunkMetaData.getPath().equals(ColumnPath.get(columnDescriptor.getPath())) &&
-                        isOnlyDictionaryEncodingPages(columnChunkMetaData.getEncodings())) {
+        ImmutableMap.Builder<ColumnDescriptor, ParquetDictionaryDescriptor> dictionaries = ImmutableMap.builder();
+        for (ColumnChunkMetaData columnMetaData : blockMetadata.getColumns()) {
+            String[] paths = columnMetaData.getPath().toArray();
+            Optional<RichColumnDescriptor> descriptor = getDescriptor(fileSchema, requestedSchema, Arrays.asList(paths));
+            if (descriptor.isPresent()) {
```

Comment:
only build dictionaries when descriptor exist, otherwise, dictionaries is empty, will scan file, dictionary predicate will not apply